tests/test_home_meter.py

```diff
@@ -0,0 +1,244 @@
+        self.strategy.area.all_markets = market_mocks
+
+        self.strategy.event_market_cycle()
+        super_method_mock.assert_called_once()
+        self.strategy.bid_update.update_and_populate_price_settings.assert_called_with(
+            self.area_mock)
+        self.strategy.offer_update.update_and_populate_price_settings.assert_called_with(
+            self.area_mock)
+        self.strategy.bid_update.reset.assert_called_with(self.strategy)
+        self.strategy.offer_update.reset.assert_called_with(self.strategy)
+
+        # We just assert calls to private methods, because we have separate unittests for them
+        self.strategy._set_energy_forecast_for_future_markets.assert_called_once_with(
+            reconfigure=False)
+        assert self.strategy._post_offer.call_count == 3
+
+        self.strategy.state.delete_past_state_values.assert_called_once_with(
+            self.area_mock.current_market.time_slot)
+        self.strategy.bid_update.delete_past_state_values.assert_called_once_with(
+            self.area_mock.current_market.time_slot)
+        self.strategy.offer_update.delete_past_state_values.assert_called_once_with(
+            self.area_mock.current_market.time_slot)
+
+    @patch("d3a_interface.constants_limits.ConstSettings.IAASettings")
+    def test_event_offer_two_sided_market(self, iaa_settings_mock):
+        """The device does not automatically react to offers in two-sided markets."""
+        iaa_settings_mock.MARKET_TYPE = 2
+        self.strategy.event_offer(market_id="some_market_id", offer="some_offer")
+        # Make sure that the method is returning immediately and not executing further logic
+        self.strategy.area.get_future_market_from_id.assert_not_called()
+
+    def test_post_offer(self):
+        """Offers are generated and sent to the offers class to be posted.
+
+        This method is private, but we test it to avoid duplication and because of its complexity.
+        """
+        self.strategy.state.get_available_energy_kWh = Mock(return_value=12)
+        market_mock = self._create_market_mocks(1)[0]
+        offer_mock = Mock()
+        market_mock.offer = Mock(return_value=offer_mock)
+        self.strategy.offer_update.initial_rate = {market_mock.time_slot: 11}
+        self.strategy.offers.post = Mock()
+
+        self.strategy._post_offer(market_mock)
+        self.strategy.offers.post.assert_called_once_with(offer_mock, market_mock.id)
+
+    def _create_profile_mock(self):
+        return OrderedDict([
+            (self.slot_times[0], 1), (self.slot_times[1], -0.5), (self.slot_times[2], -0.1)])
+
+    def _create_market_mocks(self, num_of_markets=3):
+        market_mocks = [create_autospec(OneSidedMarket) for _ in range(num_of_markets)]
+        slot_time = self.slot_times[0]
+        for market_mock in market_mocks:
+            market_mock.time_slot = slot_time
+            market_mock.id = uuid.uuid4()
+            slot_time += duration(minutes=15)
```

Comment:
Sorry, I meant to reuse it even more. Again, minor!
```suggestion
    def _create_market_mocks(self):
        market_mocks = [create_autospec(OneSidedMarket) for _ in range(self.slot_times)]
        for ii, market_mock in enumerate(market_mocks):
            market_mock.time_slot = self.slot_times[ii]
            market_mock.id = uuid.uuid4()
```
I just realise that you increase the slot_time after assigning it to market_mock.time_slot. Is this working due to the fact that is is a reference (pointer)?

Reply:
Hmm in Python it's not so straight-forward, so I hope I'll get this right. 
`slot_time = self.slot_times[0]` creates a new variable (`slot_time`) with the same reference as `self.slot_times[0]` (so to they refer to the same `datetime` object). When `slot_time += duration(minutes=15)` is executed, the reference of `slot_time` changes, but not the reference of `self.slot_times[0]`. So `self.slot_times[0]` does not change.
Python is neither pass-by-reference nor pass-by-value. When we re-assign `slot_time`, we are rebinding only its reference to another object (but not `self.slot_times[0]`'s reference). So the two references will point to two different objects.

Reply:
Talking about the specific case here: I wanted to have a flexible implementation of `_create_market_mocks` to create as many mocks as I wanted, unrelated to the number of time slots that I defined previously. That's why the time is changed dynamically (just adding 15 minutes every time). If you want, I can definitely add your suggested changes, since they wouldn't change the outcome of the current tests.

Reply:
Yeah, this is valid if you want to reuse the _create_market_mocks for a different test later on. So leave it be.